File: go/verifier/conformance_crypto_test.go
package verifier

import (
	"encoding/json"
	"os"
	"path/filepath"
	"sort"
	"testing"
)

type cryptoVector struct {
	ID    string                    `json:"id"`
	Level int                       `json:"level"`
	Title string                    `json:"title"`
	Now   string                    `json:"now"`
	Keys  map[string]map[string]JWK `json:"keys"`
	Case  *struct {
		Kind      string         `json:"kind"`
		Tokens    []string       `json:"tokens"`
		Proof     string         `json:"proof"`
		AgentKeys map[string]JWK `json:"agent_keys"`
		Htm, Htu  string
		BodyHash  string `json:"body_hash"`
		Token     string `json:"token"`
	} `json:"case"`
	World *struct {
		vectorWorld
		Peers []struct {
			EntityID   string `json:"entity_id"`
			Name       string `json:"name"`
			TrustLevel int    `json:"trust_level"`
		} `json:"peers"`
	} `json:"world"`
	Request map[string]any  `json:"request"`
	Expect  json.RawMessage `json:"expect"`
}

func TestConformanceCryptoVectors(t *testing.T) {
	files, _ := filepath.Glob(filepath.Join("..", "..", "conformance", "crypto", "*.json"))
	if len(files) < 17 {
		t.Fatalf("expected the level 2/3 vector set, found %d", len(files))
	}
	sort.Strings(files)
	for _, f := range files {
		raw, _ := os.ReadFile(f)
		var v cryptoVector
		if err := json.Unmarshal(raw, &v); err != nil {
			t.Fatalf("%s: %v", f, err)
		}
		t.Run(v.ID+" "+v.Title, func(t *testing.T) {
			now := ParseISO(v.Now)
			resolve := func(kid, iss string) *JWK {
				if k, ok := v.Keys[iss][kid]; ok {
					return &k
				}
				return nil
			}
			var e map[string]any
			_ = json.Unmarshal(v.Expect, &e)
			want := func(k string) any { return e[k] }
			if v.Case != nil {
				switch v.Case.Kind {
				case "delegation_chain":
					r := VerifyDelegationChain(v.Case.Tokens, resolve, now)
					if r.OK != want("ok").(bool) {
						t.Fatalf("ok=%v reason=%s at=%d", r.OK, r.Reason, r.At)
					}
					if r.OK {
						wc, _ := json.Marshal(want("chain"))
						gc, _ := json.Marshal(r.Chain)
						if string(wc) != string(gc) || r.Subject != want("subject").(string) {
							t.Fatalf("chain %s subject %s", gc, r.Subject)
						}
					} else if r.Reason != want("reason").(string) || float64(r.At) != want("at").(float64) {
						t.Fatalf("reason %s at %d", r.Reason, r.At)
					}
				case "request_proof":
					r := VerifyRequestProof(ProofInput{Proof: v.Case.Proof, ResolveKey: func(kid string) *JWK {
						if k, ok := v.Case.AgentKeys[kid]; ok {
							return &k
						}
						return nil
					}, Htm: v.Case.Htm, Htu: v.Case.Htu, BodyHash: v.Case.BodyHash, Now: now})
					if r.OK != want("ok").(bool) {
						t.Fatalf("ok=%v reason=%s", r.OK, r.Reason)
					}
					if r.OK {
						if r.Claims.Iss != want("iss").(string) || r.Alg != want("alg").(string) {
							t.Fatalf("iss %s alg %s", r.Claims.Iss, r.Alg)
						}
					} else if r.Reason != want("reason").(string) {
						t.Fatalf("reason %s", r.Reason)
					}
				case "attestation":
					p, left, reason := VerifyAttestation(v.Case.Token, resolve, now)
					if (reason == "") != want("ok").(bool) {
						t.Fatalf("reason=%q", reason)
					}
					if reason == "" {
						if p.Sub != want("sub").(string) || p.Atp.Action != want("action").(string) || float64(left) != want("expires_in").(float64) {
							t.Fatalf("sub %s action %s left %d", p.Sub, p.Atp.Action, left)
						}
					} else if reason != want("reason").(string) {
						t.Fatalf("reason %s", reason)
					}
				}
				return
			}
			// level 3
			deps := depsFrom(v.World.vectorWorld)
			deps.GetOrgKey = resolve
			deps.GetFederatedIssuer = func(iss string) *FederatedIssuer {
				for _, p := range v.World.Peers {
					if p.EntityID == iss {
						return &FederatedIssuer{Name: p.Name, TrustLevel: p.TrustLevel}
					}
				}
				return nil
			}
			req, bad := ParseVerifyRequest(v.Request)
			if bad != "" {
				t.Fatal(bad)
			}
			r := Verify(req, deps, Options{Now: now})
			if r.Decision != want("decision").(string) {
				t.Fatalf("decision %s (%v)", r.Decision, r.Reasons)
			}
			if inc, ok := want("reasons_include").([]any); ok {
				for _, x := range inc {
					found := false
					for _, rr := range r.Reasons {
						if rr == x.(string) {
							found = true
						}
					}
					if !found {
						t.Errorf("missing reason %v in %v", x, r.Reasons)
					}
				}
			}
			if ev, ok := want("evidence").(map[string]any); ok {
				for step, status := range ev {
					got := ""
					for _, x := range r.Evidence {
						if x.Step == step {
							got = x.Status
							break
						}
					}
					if got != status.(string) {
						t.Errorf("step %s: %q want %q", step, got, status)
					}
				}
			}
		})
	}
}

File: packages/verifier/tests/conformance.crypto.test.ts
import { describe, it, expect } from "vitest";
import { resourceContains } from "../src/delegation/attenuate";
import { readFileSync, readdirSync } from "node:fs";
import { join } from "node:path";
import { verifyDelegationChain, verifyRequestProof, verifyAttestation, verifyApprovalJwt, verify, type VerifierDeps, type AgentInfo, type CredentialInfo, type PrincipalInfo, type AttestationInfo, type ActivePolicy } from "../src/index";
import type { Delegation, CapSet, VerifyRequest } from "../src/types";

interface CryptoVector { id: string; level: number; title: string; now: string; keys: Record<string, Record<string, JsonWebKey>>; case?: Record<string, unknown> & { kind: string }; world?: { agent: AgentInfo | null; credential: CredentialInfo | null; principals: PrincipalInfo[]; delegations: Delegation[]; capabilities: CapSet; revoked: string[]; attestations: AttestationInfo[]; policy: ActivePolicy | null; peers?: { entity_id: string; name: string; trust_level: 1 | 2; stale_ok_seconds?: number }[]; consumed?: string[]; peer_status?: Record<string, "active" | "revoked" | "unreachable"> }; request_hash?: string; audience?: string; request?: VerifyRequest; expect: Record<string, unknown> }

const dir = join(process.cwd(), "conformance", "crypto");
const vectors: CryptoVector[] = readdirSync(dir).filter((f) => f.endsWith(".json")).sort().map((f) => JSON.parse(readFileSync(join(dir, f), "utf8")));
/** v0.2 single-use: a consumption store pre-seeded with the vector's `consumed` jtis. */
const consumeStore = (seed: string[]) => { const seen = new Set(seed); return async (jti: string) => { if (seen.has(jti)) return "seen" as const; seen.add(jti); return "first" as const; }; };
const resolver = (keys: CryptoVector["keys"]) => async (kid: string, iss: string) => keys[iss]?.[kid] ?? null;

describe("conformance levels 2–3 (signed objects, federation)", () => {
  it("has the reference set", () => { expect(vectors.filter((v) => v.level === 2).length).toBeGreaterThanOrEqual(12); expect(vectors.filter((v) => v.level === 3).length).toBeGreaterThanOrEqual(5); });
  for (const v of vectors) {
    it(`${v.id} ${v.title}`, async () => {
      const now = new Date(v.now); const e = v.expect;
      if (v.case?.kind === "delegation_chain") { const r = await verifyDelegationChain(v.case.tokens as string[], resolver(v.keys), now); expect(r.ok).toBe(e.ok); if (r.ok) { expect(r.chain).toEqual(e.chain); expect(r.subject).toBe(e.subject); expect(r.effective).toEqual(e.effective); } else { expect(r).toMatchObject({ at: e.at, reason: e.reason }); } return; }
      if (v.case?.kind === "request_proof") { const c = v.case as unknown as { proof: string; agent_keys: Record<string, JsonWebKey>; htm: string; htu: string; body_hash: string }; const r = await verifyRequestProof({ proof: c.proof, resolveKey: async (kid) => (kid ? c.agent_keys[kid] ?? null : null), htm: c.htm, htu: c.htu, bodyHash: c.body_hash, now }); expect(r.ok).toBe(e.ok); if (r.ok) { expect(r.claims.iss).toBe(e.iss); expect(r.alg).toBe(e.alg); } else expect(r.reason).toBe(e.reason); return; }
      if (v.case?.kind === "approval") { const r = await verifyApprovalJwt(v.case.token as string, resolver(v.keys), now); expect(r.ok).toBe(e.ok); if (r.ok) { expect(r.header.kid).toBe(e.kid); expect(r.payload.jti).toBe(e.jti); expect(r.payload.atp.status).toBe(e.status); expect(r.payload.atp.device ?? null).toBe(e.device ?? null); expect(r.payload.atp.device_proof_hash ?? null).toBe(e.device_proof_hash ?? null); } else expect(r.reason).toBe(e.reason); return; }
      if (v.case?.kind === "attestation") { const r = await verifyAttestation(v.case.token as string, resolver(v.keys), now); expect(r.ok).toBe(e.ok); if (r.ok) { expect(r.payload.sub).toBe(e.sub); expect(r.payload.atp.action).toBe(e.action); expect(r.expires_in).toBe(e.expires_in); if (e.use !== undefined) expect(r.payload.atp.use).toBe(e.use); } else expect(r.reason).toBe(e.reason); return; }
      // level 3: full pipeline with peers
      const w = v.world!; const revoked = new Set(w.revoked);
      const deps: VerifierDeps = {
        getAgent: async (ref) => (w.agent && (ref === w.agent.id || ref === w.agent.stable_id) ? w.agent : null),
        getActiveCredential: async () => w.credential, getPrincipal: async (id) => w.principals.find((p) => p.id === id) ?? null,
        getLeafDelegation: async (agentId, delegationId) => w.delegations.find((d) => (delegationId ? d.id === delegationId : d.subject_agent_id === agentId)) ?? null,
        getDelegationAncestry: async () => w.delegations, getAgentCapabilities: async () => w.capabilities,
    getCoveringDelegation: async (agentId, action, resource) => w.delegations.find((d) => d.subject_agent_id === agentId && d.status === "active" && d.capabilities.some((c) => c.action === action && (!resource || resourceContains(c.resource, resource)))) ?? null,
 isRevoked: async (_t, id) => revoked.has(id), getAttestations: async () => w.attestations, getPolicy: async () => w.policy,
        ...(w.consumed ? { consumeAttestation: consumeStore(w.consumed) } : {}),
        ...(w.peer_status ? { peerAttestationStatus: async (_iss: string, jti: string) => w.peer_status![jti] ?? "active" } : {}),
        getOrgKey: resolver(v.keys), getFederatedIssuer: async (iss) => { const p = (w.peers ?? []).find((x) => x.entity_id === iss); return p ? { name: p.name, trust_level: p.trust_level, stale_ok_seconds: p.stale_ok_seconds } : null; },
      };
      const r = await verify(v.request!, deps, { now, requestHash: v.request_hash, audience: v.audience });
      expect(r.decision).toBe(e.decision);
      for (const reason of (e.reasons_include as string[]) ?? []) expect(r.reasons).toContain(reason);
      for (const [step, status] of Object.entries((e.evidence as Record<string, string>) ?? {})) expect(r.evidence.find((x) => x.step === step)?.status, `${v.id} step ${step}`).toBe(status);
    });
  }
});
